File: python/agent/hacker_news_agent/workflows.py
from typing import Any, Dict, List, Optional

from dbos import DBOS

from .agent import (evaluate_results_step, generate_follow_ups_step,
                    plan_research_step, should_continue_step)
from .api import get_comments_step, search_hackernews_step
from .llm import synthesize_findings_step


@DBOS.step()
def log_research_step(message: str) -> None:
    """Log research progress messages."""
    from rich.console import Console
    console = Console()
    console.print(f"[dim]{message}[/dim]")


@DBOS.workflow()
def research_iteration_workflow(
    topic: str, query: str, iteration: int, include_comments: bool = True
) -> Dict[str, Any]:
    """Execute a single research iteration."""

    log_research_step(f"🔍 Searching for stories: '{query}'")
    
    # Search for stories
    stories = search_hackernews_step(query, max_results=30)
    
    if stories:
        log_research_step(f"📚 Found {len(stories)} stories, analyzing diverse selection...")
        
        # Log interesting stories we're analyzing - show variety, not just top ones
        for i, story in enumerate(stories[:8]):  # Show more stories
            title = story.get("title", "No title")[:60]
            points = story.get("points", 0)
            num_comments = story.get("num_comments", 0)
            log_research_step(f"  📖 Story {i+1}: {title}... ({points} points, {num_comments} comments)")
    else:
        log_research_step("❌ No stories found for this query")

    # Optionally get comments for diverse stories
    comments = []
    if include_comments and stories:
        # Get comments from diverse stories (mix of high and medium engagement)
        # Take top 2 stories and 2 from middle range for diversity
        high_engagement = [
            story for story in stories[:3] 
            if story.get("points", 0) > 5 and story.get("num_comments", 0) > 3
        ][:2]
        
        medium_engagement = [
            story for story in stories[3:10] 
            if story.get("points", 0) > 0 and story.get("num_comments", 0) > 1
        ][:2]
        
        selected_stories = high_engagement + medium_engagement

        if selected_stories:
            log_research_step(f"💬 Reading comments from {len(selected_stories)} diverse stories...")
            
            for i, story in enumerate(selected_stories):
                story_id = story.get("objectID")
                title = story.get("title", "Unknown")[:40]
                if story_id:
                    log_research_step(f"  💭 Reading comments from: {title}...")
                    story_comments = get_comments_step(story_id, max_comments=15)
                    comments.extend(story_comments)
                    log_research_step(f"    ✓ Read {len(story_comments)} comments")

    # Evaluate results
    log_research_step(f"🤔 Analyzing findings from {len(stories)} stories and {len(comments)} comments...")
    evaluation = evaluate_results_step(topic, query, stories, comments)

    return {
        "iteration": iteration,
        "query": query,
        "stories_found": len(stories),
        "comments_analyzed": len(comments),
        "evaluation": evaluation,
        "stories": stories,
        "comments": comments,
    }


@DBOS.workflow()
def agentic_research_workflow(
    topic: str, max_iterations: Optional[int] = None, include_comments: bool = True
) -> Dict[str, Any]:
    """Main agentic research workflow that autonomously researches a topic."""

    # Step 1: Agent creates initial research plan (just for context, not all queries)
    log_research_step(f"🎯 Agent planning research approach for: {topic}")
    research_plan = plan_research_step(topic)

    # Use provided max_iterations or default to 8
    if max_iterations is None:
        max_iterations = 8

    # Initialize tracking
    all_findings = []
    research_history = []
    current_iteration = 0

    # Step 2: Start with the original topic as first query
    current_query = topic
    
    # Main iterative research loop
    while current_iteration < max_iterations:
        current_iteration += 1
        
        log_research_step(f"🔄 Starting iteration {current_iteration}/{max_iterations}")
        
        # Execute research iteration
        iteration_result = research_iteration_workflow(
            topic, current_query, current_iteration, include_comments
        )
        research_history.append(iteration_result)
        all_findings.append(iteration_result["evaluation"])
        
        # Check if we found anything useful
        stories_found = iteration_result["stories_found"]
        relevance_score = iteration_result["evaluation"].get("relevance_score", 0)
        
        if stories_found == 0:
            log_research_step(f"⚠️  No stories found for '{current_query}', trying alternative approach...")
            
            # Generate alternative queries when we find nothing
            alternative_queries = generate_follow_ups_step(
                topic, all_findings, current_iteration
            )
            
            if alternative_queries:
                # Try the first alternative
                current_query = alternative_queries[0]
                log_research_step(f"🔄 Retrying with: '{current_query}'")
                continue
            else:
                log_research_step("❌ No alternative queries available, continuing...")
        
        # If we have enough iterations, check if we should continue
        if current_iteration >= 2:
            log_research_step("🤔 Agent evaluating whether to continue research...")
            decision = should_continue_step(
                topic, all_findings, current_iteration, max_iterations
            )

            if not decision.get("should_continue", False):
                log_research_step(f"✅ Agent decided to conclude research: {decision.get('reason', 'No reason provided')}")
                break

        # Generate next research question based on current findings
        if current_iteration < max_iterations:
            log_research_step("💭 Agent generating next research question...")
            follow_up_queries = generate_follow_ups_step(
                topic, all_findings, current_iteration
            )

            if follow_up_queries:
                current_query = follow_up_queries[0]  # Take the first suggestion
                log_research_step(f"➡️  Next research focus: '{current_query}'")
            else:
                log_research_step("💡 No new research directions found, concluding...")
                break

    # Step 3: Agent synthesizes final report
    log_research_step("📋 Agent synthesizing final research report...")
    final_report = synthesize_findings_step(topic, all_findings)

    return {
        "topic": topic,
        "research_plan": research_plan,
        "total_iterations": current_iteration,
        "max_iterations": max_iterations,
        "research_history": research_history,
        "final_report": final_report,
        "summary": {
            "total_stories": sum(r["stories_found"] for r in research_history),
            "total_comments": sum(r["comments_analyzed"] for r in research_history),
            "queries_executed": [r["query"] for r in research_history],
            "avg_relevance": (
                sum(f.get("relevance_score", 0) for f in all_findings)
                / len(all_findings)
                if all_findings
                else 0
            ),
        },
    }

File: typescript/hacker-news-agent/src/workflows.ts
import { DBOS } from "@dbos-inc/dbos-sdk";
import { evaluateResults, generateFollowUps, shouldContinue } from "./agent";
import { searchHackerNews, getComments } from "./api";
import { synthesizeFindings, Finding } from "./llm";

export interface IterationResult {
  iteration: number;
  query: string;
  stories_found: number;
  comments_analyzed: number;
  evaluation: any;
  stories: any[];
  comments: any[];
}

export interface ResearchResult {
  topic: string;
  total_iterations: number;
  max_iterations: number;
  research_history: IterationResult[];
  final_report: any;
  summary: {
    total_stories: number;
    total_comments: number;
    queries_executed: string[];
    avg_relevance: number;
  };
}

async function agenticResearchWorkflowFunction(
  topic: string,
  maxIterations: number,
): Promise<ResearchResult> {
  console.log(`Starting agentic research for: ${topic}`);

  const allFindings: Finding[] = [];
  const researchHistory: IterationResult[] = [];
  let currentIteration = 0;
  let currentQuery = topic;

  // Main agentic research loop
  while (currentIteration < maxIterations) {
    currentIteration++;
    console.log(`🔄 Starting iteration ${currentIteration}/${maxIterations}`);

    // Research the next query
    const iterationResult = await researchQueryWorkflow(
      topic,
      currentQuery,
      currentIteration,
    );
    researchHistory.push(iterationResult);
    allFindings.push(iterationResult.evaluation);

    // Handle cases where no results are found
    const storiesFound = iterationResult.stories_found;
    if (storiesFound === 0) {
      console.log(
        `⚠️  No stories found for '${currentQuery}', trying alternative approach...`,
      );

      // Generate alternative queries when hitting dead ends
      const alternativeQuery = await DBOS.runStep(
        () => generateFollowUps(topic, allFindings, currentIteration),
        { name: "generateFollowUps" },
      );
      if (alternativeQuery) {
        currentQuery = alternativeQuery;
        console.log(`🔄 Retrying with: '${currentQuery}'`);
        continue;
      } else {
        console.log("❌ No alternative queries available, continuing...");
      }
    }

    // Evaluate whether to continue research
    console.log("🤔 Agent evaluating whether to continue research...");
    const shouldContinueDecision = await DBOS.runStep(
      () => shouldContinue(topic, allFindings, currentIteration, maxIterations),
      { name: "shouldContinue" },
    );
    if (!shouldContinueDecision) {
      console.log("✅ Agent decided to conclude research");
      break;
    }

    // Generate next research question based on findings
    if (currentIteration < maxIterations) {
      console.log("💭 Agent generating next research question...");
      const followUpQuery = await DBOS.runStep(
        () => generateFollowUps(topic, allFindings, currentIteration),
        { name: "generateFollowUps" },
      );
      if (followUpQuery) {
        currentQuery = followUpQuery;
        console.log(`➡️  Next research focus: '${currentQuery}'`);
      } else {
        console.log("💡 No new research directions found, concluding...");
        break;
      }
    }
  }

  // Final step: Synthesize all findings into comprehensive report
  console.log("📋 Agent synthesizing final research report...");
  const finalReport = await DBOS.runStep(
    () => synthesizeFindings(topic, allFindings),
    { name: "synthesizeFindings" },
  );

  // Return complete research results
  return {
    topic,
    total_iterations: currentIteration,
    max_iterations: maxIterations,
    research_history: researchHistory,
    final_report: finalReport,
    summary: {
      total_stories: researchHistory.reduce(
        (sum, r) => sum + r.stories_found,
        0,
      ),
      total_comments: researchHistory.reduce(
        (sum, r) => sum + r.comments_analyzed,
        0,
      ),
      queries_executed: researchHistory.map((r) => r.query),
      avg_relevance:
        allFindings.length > 0
          ? allFindings.reduce((sum, f) => sum + (f.relevance_score || 0), 0) /
            allFindings.length
          : 0,
    },
  };
}

async function researchQueryWorkflowFunction(
  topic: string,
  query: string,
  iteration: number,
): Promise<IterationResult> {
  console.log(`🔍 Searching for stories: '${query}'`);

  // Step 1: Search Hacker News for stories about the topic
  const stories = await DBOS.runStep(() => searchHackerNews(query, 30), {
    name: "searchHackerNews",
  });

  if (stories.length > 0) {
    console.log(`📚 Found ${stories.length} stories, analyzing all stories...`);
    stories.forEach((story, i) => {
      const title = (story.title || "No title").slice(0, 80);
      const points = story.points || 0;
      const numComments = story.num_comments || 0;
      console.log(
        `  📖 Story ${i + 1}: ${title}... (${points} points, ${numComments} comments)`,
      );
    });
  } else {
    console.log("❌ No stories found for this query");
  }

  // Step 2: Gather comments from all stories found
  const comments: any[] = [];
  if (stories.length > 0) {
    console.log(`💬 Reading comments from ALL ${stories.length} stories...`);

    for (let i = 0; i < stories.length; i++) {
      const story = stories[i];
      const storyId = story.objectID;
      const title = (story.title || "Unknown").slice(0, 50);
      const numComments = story.num_comments || 0;

      if (storyId && numComments > 0) {
        console.log(
          `  💭 Reading comments from: ${title}... (${numComments} comments)`,
        );
        const storyComments = await DBOS.runStep(
          () => getComments(storyId, 10),
          { name: "getComments" },
        );
        comments.push(...storyComments);
        console.log(`    ✓ Read ${storyComments.length} comments`);
      } else if (storyId) {
        console.log(`  📖 Story has no comments: ${title}`);
      } else {
        console.log(`  ❌ No story ID available for: ${title}`);
      }
    }
  }

  // Step 3: Evaluate gathered data and return findings
  console.log(
    `🤔 Analyzing findings from ${stories.length} stories and ${comments.length} comments...`,
  );
  const evaluation = await DBOS.runStep(
    () => evaluateResults(topic, query, stories, comments),
    { name: "evaluateResults" },
  );

  return {
    iteration,
    query,
    stories_found: stories.length,
    comments_analyzed: comments.length,
    evaluation,
    stories,
    comments,
  };
}

// Register DBOS workflows
export const agenticResearchWorkflow = DBOS.registerWorkflow(
  agenticResearchWorkflowFunction,
  { name: "agenticResearchWorkflow" },
);
export const researchQueryWorkflow = DBOS.registerWorkflow(
  researchQueryWorkflowFunction,
  { name: "researchQueryWorkflow" },
);
